benchmarks/distributed/rpc/rl/agent.py

```diff
@@ -0,0 +1,177 @@
+        from observer import ObserverBase
+
+        self.batch = batch
+        self.policy = Policy(state_size, nlayers, out_features, self.batch)
+        self.optimizer = optim.Adam(self.policy.parameters(), lr=1e-2)
+
+        self.batch_size = batch_size
+        for rank in range(batch_size):
+            ob_info = rpc.get_worker_info(OBSERVER_NAME.format(rank + 2))
+
+            self.rewards[ob_info.id] = []
+
+        self.saved_log_probs = [] if self.batch else {
+            k: [] for k in range(self.batch_size)}
+
+        self.pending_states = self.batch_size
+        self.state_size = state_size
+        self.states = torch.zeros(self.batch_size, *state_size)
+
+    def store_ob_rrefs(self, ob_rrefs):
+        for ob_rref in ob_rrefs:
+            self.ob_rrefs.append(ob_rref)
+
+
+    @staticmethod
+    @rpc.functions.async_execution
+    def select_action_batch(agent_rref, observer_id, state):
+        self = agent_rref.local_value()
+        observer_id -= 2
+
+
+
+        self.states[observer_id].copy_(state)
+        future_action = self.future_actions.then(
+            lambda future_actions: future_actions.wait()[observer_id].item()
+        )
+
+        with self.lock:
+            if self.pending_states == self.batch_size:
+                self.agent_latency_start = time.time()
+            self.pending_states -= 1
+            if self.pending_states == 0:
+                self.pending_states = self.batch_size
+                probs = self.policy(self.states)
+                m = Categorical(probs)
+                actions = m.sample()
+                self.saved_log_probs.append(m.log_prob(actions).t())
+                future_actions = self.future_actions
+                self.future_actions = torch.futures.Future()
+                future_actions.set_result(actions)
+
+                self.agent_latency_end = time.time()
+                # agent_latency_end = time.time()
+
+                batch_latency = self.agent_latency_end - self.agent_latency_start
+                # batch_latency = agent_latency_end - agent_latency_start
+                self.agent_latency.append(batch_latency)
+                self.agent_throughput.append(self.batch_size / batch_latency)
+
+        return future_action
+
+    @staticmethod
+    def select_action_non_batch(agent_rref, observer_id, state):
+        self = agent_rref.local_value()
+        observer_id -= 2
+
+        agent_latency_start = time.time()
+
+        state = state.float().unsqueeze(0)
+        probs = self.policy(state)
+        m = Categorical(probs)
+        action = m.sample()
+
+        self.saved_log_probs[observer_id].append(m.log_prob(action))
+
+        agent_latency_end = time.time()
+
+        # non_batch_latency = self.agent_latency_end - self.agent_latency_start
+        non_batch_latency = agent_latency_end - agent_latency_start
+
+        self.agent_latency.append(non_batch_latency)
+        self.agent_throughput.append(1 / non_batch_latency)
```

Comment:
there can be race on these two lines. As there are multiple observers accessing this function, it is not guaranteed these two lines will always run atomically. You might need a lock here as well.